src/expr/src/scalar/mod.rs

```diff
@@ -328,30 +328,51 @@ impl MirScalarExpr {
     }
 
     /// For a given `expr`, if `self` is `<expr> = <literal>` or `<literal> = <expr>` then
-    /// return `<literal>`.
-    pub fn expr_eq_literal(&self, expr: &MirScalarExpr) -> Option<Datum> {
+    /// return `(<literal>, false)`. In addition to just trying to match `<expr>` as it is, it also
+    /// tries to remove an invertible function call (such as a cast). If the match match succeeds
+    /// with the inversion, then it returns `(<inverted-literal>, true)`. For more details on the
+    /// inversion, see `invert_casts_on_expr_eq_literal_inner`.
+    pub fn expr_eq_literal(&self, expr: &MirScalarExpr) -> Option<(Row, bool)> {
         if let MirScalarExpr::CallBinary {
             func: BinaryFunc::Eq,
             expr1,
             expr2,
         } = self
         {
-            if let Some(Ok(datum1)) = expr1.as_literal() {
-                if &**expr2 == expr {
-                    return Some(datum1);
-                }
+            if let Some(Ok(lit)) = expr1.as_literal_owned() {
+                return Self::expr_eq_literal_inner(expr, lit, expr1, expr2);
+            }
+            if let Some(Ok(lit)) = expr2.as_literal_owned() {
+                return Self::expr_eq_literal_inner(expr, lit, expr2, expr1);
             }
-            if let Some(Ok(datum2)) = expr2.as_literal() {
-                if &**expr1 == expr {
-                    return Some(datum2);
+        }
+        None
+    }
+
+    fn expr_eq_literal_inner(
```

Comment:
I thought this was moving completely to `canonicalize_mfp`. What's the reason of keeping this method?

Reply:
This is the join issue: https://github.com/MaterializeInc/materialize/issues/4171

Reply:
Got it.